src/Map/Map.jl

```diff
@@ -25,8 +25,12 @@ end
 
 function preimage(M::Map{D, C}, a) where {D, C}
   if isdefined(M.header, :preimage)
-    p = M.header.preimage(a)::elem_type(D)
-    @assert parent(p) === domain(M)
+    if isdefined(M.header, :domain)
+      p = M.header.preimage(a)::elem_type(M.header.domain)
```

Comment:
Can you elaborate why this change is necessary? It also leads to failing tests in Oscar.

Since we assert `parent(p) === domain(M)`, what is wrong with doing `::elem_type(D)`?

Reply:
I made the same changes as for `image`... the problem that was occurring is that for some elements, the type `Map` is provided with more than 2 parameters, so when we ask for `elem_type(C)` for `Map{D, C}`, this was crashing because `C` was a tuple of types, not only a type by itself. So I had to change it to call directly the (co)domain, and not the parameters which was not always a type (in my case)... Then I adapted a previous code which was using a type `Map` with 4 parameters, but I can restrict it to 2 and avoid the last 2 parameters if it really breaks everything..

Reply:
This looks very wrong. For which `Map{D, C}` is `C` is not (a supertype) of the type of the codomain?

Reply:
(The map interface requires that `D` or `C` should be (supertypes of) the type of domain/codomain. If a map implementation is violating this, then of course everything stops working. So it would be good to know which map implementations violate the interface.)

Reply:
At least the one I have added, and the one I have adapted. So in the file in which I have added the `NfRelOrdToRelFinFieldMor`. I can make a new PR where I change this and put back the old versions of `image` and `preimage` if you want

Reply:
Or I am mistaking with `MapHeader`... but as far as I see, in every files of `Map.jl`, every structures have as a supertype `Map` with 4 parameters.

Reply:
(I am currently running the tests with the old version to see what goes boom)

Reply:
I can't reproduce the bug that made me make those changes... I may have fixed it without noticing..

Reply:
So I have a fix: I put back the old definitions of `image` and `preimage` and my codes/tests running (at least on my laptop)

Reply:
Yes, `Map` has four parameters, but the first two have the meaning as I described above. If one writes `Map{C, D}`, then this is equivalent to `Map{C, D, <: Any, <: Any}`.